crates/esthri/src/ops/sync.rs

```diff
@@ -169,77 +170,86 @@ fn process_globs<'a, P: AsRef<Path> + 'a>(path: P, filters: &[GlobFilter]) -> Op
     }
 }
 
-fn create_dirent_stream<'a>(
-    directory: &'a Path,
-    filters: &'a [GlobFilter],
-) -> impl Stream<Item = Result<(String, Option<ListingMetadata>)>> + 'a {
-    async_stream::stream! {
-        for entry in WalkDir::new(directory) {
-            let entry = if let Ok(entry) = entry {
-                entry
-            } else {
-                yield Err(entry.err().unwrap().into());
-                return;
-            };
-            if entry.file_name().to_string_lossy().contains(TEMP_FILE_PREFIX) {
-                continue;
-            }
-            let metadata = entry.metadata();
-            let stat = if let Ok(stat) = metadata {
-                stat
-            } else {
-                yield Err(metadata.err().unwrap().into());
-                return;
-            };
-            if stat.is_dir() {
-                continue;
-            }
-            if entry.path_is_symlink() {
-                warn!("symlinks are ignored");
-                continue;
-            }
-            let path = entry.path();
-            debug!("local path={}", path.display());
-            if process_globs(&path, filters).is_some() {
-                yield Ok((path.to_string_lossy().into(), ListingMetadata::none()));
+fn read_dir(directory: &Path, filters: &[GlobFilter]) -> impl Stream<Item = Result<PathBuf>> {
+    let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+    let walk_dir = WalkDir::new(directory);
+    let filters = filters.to_vec();
+    tokio::task::spawn_blocking(move || async move {
+        for entry in walk_dir {
+            match entry {
+                Ok(entry) => {
+                    if entry
+                        .file_name()
+                        .to_string_lossy()
+                        .contains(TEMP_FILE_PREFIX)
+                    {
+                        continue;
+                    }
+                    if entry.path_is_symlink() {
+                        warn!("symlinks are ignored");
+                        continue;
+                    }
+                    debug!("local path={}", entry.path().display());
+                    if process_globs(entry.path(), &filters).is_some() {
+                        tx.send(Ok(entry.into_path()))
+                            .expect("failed to send file path");
+                    }
+                }
+                Err(e) => {
+                    tx.send(Err(e.into())).expect("failed to send error");
+                    break;
+                }
             }
         }
-    }
+    });
+    UnboundedReceiverStream::new(rx)
```

Comment:
Why do we need to spawn a thread in order to generate an async list of paths?

Reply:
WalkDir is built on top of [ReadDir](https://doc.rust-lang.org/stable/std/fs/struct.ReadDir.html) which is blocking, so we either need an async WalkDir (tokio has an async version of ReadDir but we'd have to implement the recursive bit) or we need to consume the iterator inside a call to spawn_blocking

Reply:
And it might not actually spawn a thread if there is an idle blocking thread ready

Reply:
Nice

Reply:
Just for my own understanding, does that mean that previously when we did

```
let dirent_stream = create_dirent_stream(directory, filters);
let etag_stream = translate_paths(dirent_stream, compression).buffer_unordered(task_count);
```

The whole stream of futures might've been blocked while it waited on WalkDir? And practically, maybe that was okay-ish because it probably would've never blocked for long?

Reply:
Each iteration of the `WalkDir` loop had a potential block (probably by making a call to read_dir). I think you're right in that it's not a big problem because it's all metadata calls not actual io.

> The whole stream of futures might've been blocked while it waited on WalkDir?

It kind of depends on the runtime. If you call a blocking operation in a non-blocking thread you block the runtime that you are on. In the case of a multi-threaded runtime you might not even notice because the other threads will be unaffected. 

I don't know exactly where the blocking would happen in this case with all the nested streams/buffer_unordereds by generally it'd look like something like this -

```rust
async fn f() -> i32 {
    let num = http_call().await;
    num
}

async fn bad_f() -> i32 {
    let num = http_call().await;
    blocking_call();
    num
}

async fn g() {
    // iterator of 5 futures
    let reqs = std::iter::from_fn(|| f()).take(5);
    let reqs = futures::stream::iter(reqs);
    // await them all at the same time
    reqs.buffer_unordered(5).await;
}
```

if (assuming single threaded runtime) you swapped f and bad_f you'd run into a situation where future-1 got a response from `http_call` and started `blocking_call`. Now futures 2-5 will not be able to make progress on their `http_call`s because the thread is stuck awaiting on `blocking_call`.

The tldr is that async rust is cooperative multitasking. `.await` is how you yield to the scheduler so it can make progress on other tasks. If the gap between awaits is too long then you risk starving whatever else is going on in the background of cpu time

Reply:
> Each iteration of the `WalkDir` loop had a potential block (probably by making a call to read_dir). I think you're right in that it's not a big problem because it's all metadata calls not actual io.

My understanding is that it is actual IO, because read_dir would need to make call to the filesystem, which could end up reading from disk (if things aren't already cached somewhere in the kernel) to pull in the directory inode data structure.  The read of the inode is likely small and fast, but it is IO that can potentially block for a long time (if it's a slow drive, or a network drive, etc).